### PR TITLE
Fix division without bcmath and using the precision parameter

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Introduce phpstan, github actions, php-cs-fixer and also fix a bug when using a precision without having bcmath set. Also update dependencies and increase the PHP requirement to PHP7.2+. <br> The actual fix about precision: https://github.com/PrestaShop/decimal/pull/18/commits/5d39d4eda4b5fbe08c68b45e2670c7f3fdfdec60#diff-25c0e6754f8433ec8448cb2b71ba8d0a66b6574dcb8ed25e27a87eaf28c58996R115
| Type?             | bug fix
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Ci is green
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

### CODE REVIEW

**Comment by matks** [comment]
> Introduce phpstan, github actions, php-cs-fixer and also fix a bug when using a precision without having bcmath set. Also update dependencies and increase the PHP requirement to PHP7.2+.

Maybe it would have been easier to submit different pull requests for each item 😛

**Comment by PierreRambaud** [comment]
> > Introduce phpstan, github actions, php-cs-fixer and also fix a bug when using a precision without having bcmath set. Also update dependencies and increase the PHP requirement to PHP7.2+.
> 
> Maybe it would have been easier to submit different pull requests for each item

No time, do you have time? The fix is a one line, and couldn't wait 3 years before being merge. Please rebase, review and merge fast!

File: tests/Operation/RoundingTest.php [hunk 1]
```diff
@@ -8,21 +8,22 @@
...
     /**
      * Given decimal number
      * When rounding it using an undefined rounding mode
      * An InvalidArgumentException should be thrown
-     *
-     * @expectedException \InvalidArgumentException
      */
     public function testItThrowsExceptionIfRoundingModeIsInvalid()
     {
+        $this->expectException(InvalidArgumentException::class);
```
Comment (matks) [inline tests/Operation/RoundingTest.php:25]: Agreed. I prefer assertions rather than phpDoc magic annotation

File: tests/phpstan/phpstan.neon [hunk 2]
```diff
@@ -0,0 +1,7 @@
+parameters:
+  paths:
+    - ./
+  excludes_analyse:
+    - vendor/
+  reportUnmatchedIgnoredErrors: false
+  level: max
```
Comment (matks) [inline tests/phpstan/phpstan.neon:7]: 👍

File: src/Operation/Division.php [hunk 3]
```diff
@@ -112,9 +115,7 @@ public function computeWithoutBcMath(DecimalNumber $a, DecimalNumber $b, $precis
             $b = $b->toMagnitude($maxPrecision);
         }
 
-        $result = $this->integerDivision($a, $b, max($precision, $aPrecision));
-
-        return $result;
+        return $this->integerDivision($a, $b, $precision);
```
Comment (matks) [inline src/Operation/Division.php:118]: In the current implementation
- if `$aPrecision > $precision` it uses `$aPrecision` although `$precision` have been requested by user
- if `$bPrecision > $aPrecision` it uses `$aPrecision`

I agree your code seems more valid.

**Review by matks — DISMISSED** [review]
LGTM

**Comment by matks** [comment]
> > > Introduce phpstan, github actions, php-cs-fixer and also fix a bug when using a precision without having bcmath set. Also update dependencies and increase the PHP requirement to PHP7.2+.
> > 
> > 
> > Maybe it would have been easier to submit different pull requests for each item
> 
> No time, do you have time? The fix is a one line, and couldn't wait 3 years before being merge. Please rebase, review and merge fast!

Do I even know php?

File: tests/Operation/RoundingTest.php (hunk 1 above)
Comment (PierreRambaud) [reply tests/Operation/RoundingTest.php:25]: The magic annotation has been deprecated and no longer work in the next release, that's the reason ^^

**Comment by PierreRambaud** [comment]
> > > > Introduce phpstan, github actions, php-cs-fixer and also fix a bug when using a precision without having bcmath set. Also update dependencies and increase the PHP requirement to PHP7.2+.
> > > 
> > > 
> > > Maybe it would have been easier to submit different pull requests for each item
> > 
> > 
> > No time, do you have time? The fix is a one line, and couldn't wait 3 years before being merge. Please rebase, review and merge fast!
> 
> Do I even know php?

Who need to know something someone already know better than you need to know! :eyes:

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Why do you add all Github Actions ? All have already done in `develop`

File: .github/workflows/php.yml [hunk 4]
```diff
@@ -1,46 +1,123 @@
...
+        run: ./vendor/bin/phpstan analyse -c tests/phpstan/phpstan.neon
+
+  phpunit:
+    name: PHPUnit
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        php-versions: ['5.4', '5.5', '5.6', '7.0', '7.1', '7.2', '7.3', '7.4']
-        #, '8.0', '8.1'
+        php: ['7.2', '7.3', '7.4', '8.0']
```
Comment (Progi1984) [inline .github/workflows/php.yml:81]:
```suggestion
        php: ['7.2', '7.3', '7.4', '8.0', '8.1']
```

Comment (PierreRambaud) [reply .github/workflows/php.yml:81]: Not yet

**Review by Progi1984 — DISMISSED** [review]

File: composer.json [hunk 5]
```diff
@@ -15,16 +15,16 @@
...
         }
     ],
     "require": {
-        "php" : ">=5.4"
+        "php" : ">=7.2"
     },
     "require-dev": {
-        "phpunit/phpunit" : "4.*"
+        "phpunit/phpunit" : "8.*",
+        "codacy/coverage": "dev-master",
```
Comment (matks) [inline composer.json]: Theoretically this exposes us to supply chain attacks.

File: tests/Operation/RoundingTest.php (hunk 1 above)
Comment (matks) [reply tests/Operation/RoundingTest.php:25]: I prefer the latest trends always

**Review by matks — DISMISSED** [review]

File: composer.json (hunk 5 above)
Comment (PierreRambaud) [reply composer.json]: Don't know where this one comes from

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
I approve the PR and CI is green. We could merge or maybe ask for another review from @PrestaShop/prestashop-maintainers

**Review by atomiix — APPROVED** [review]

**Comment by atomiix** [comment]
Thanks @PierreRambaud

**Pull request merged**